src/orange/orangeinp/IntersectRegion.cc

```diff
@@ -1271,6 +1264,112 @@ bool Prism::encloses(Prism const& other) const
+ * - segments parallel to z become z-aligned cylinders,
+ * - segments perpendicular to z become z-orthagonal planes,
+ * - other segments become z-aligned cones.
+ *
+ * If segment is coincident with the z axis, no surface is created, as it
+ * would enclose zero volume.
+ */
+void RevolvedSpecialTrapezoid::build(IntersectSurfaceBuilder& insert_surface) const
+{
+    auto const& bot = trap_.bot();
+    auto const& top = trap_.top();
+    constexpr auto left = Bound::lo;
+    constexpr auto right = Bound::hi;
+
+    // Create z planes first for faster short circuiting
+    if (trap_.variety() != SpecialTrapezoid::Variety::pointy_bot)
+    {
+        // Bottom z plane
+        insert_surface(Sense::outside, PlaneZ{bot.z});
+    }
+    if (trap_.variety() != SpecialTrapezoid::Variety::pointy_top)
+    {
+        // Top z plane
+        insert_surface(Sense::inside, PlaneZ{top.z});
+    }
+
+    SoftClose soft_close(trap_.abs_tol());
+
+    // Lambda for creating a cylindrical/conical surface, or no surface
+    auto make_vertical_surface
+        = [&](real_type r_bot, real_type r_top, Sense sense) {
+              if (!soft_close(r_bot, r_top))
+              {
+                  // Conical surface
+                  insert_surface(
+                      sense, this->make_cone({r_bot, bot.z}, {r_top, top.z}));
+              }
+              else if (!soft_close(real_type{0}, r_bot))
+              {
+                  // Cylindrical surface
+                  insert_surface(sense, CCylZ(r_bot));
+              }
+          };
+
+    // Create two vericle surfaces
+    make_vertical_surface(bot.r[left], top.r[left], Sense::outside);
+    make_vertical_surface(bot.r[right], top.r[right], Sense::inside);
+
+    // Establish bbox
+    auto r_max = std::max(bot.r[right], top.r[right]);
+    insert_surface(
+        Sense::inside,
+        BBox::from_unchecked({-r_max, -r_max, bot.z}, {r_max, r_max, top.z}));
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Write output to the given JSON object.
+ */
+void RevolvedSpecialTrapezoid::output(JsonPimpl* j) const
+{
+    to_json_pimpl(j, *this);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create a cone from two (r, z) points.
+ *
+ * The cone is created by revolving the line segment between the points
+ * around the z axis.
+ */
+ConeZ RevolvedSpecialTrapezoid::make_cone(Real2 p0, Real2 p1) const
+{
+    constexpr size_type R = 0;
+    constexpr size_type Z = 1;
+
+    auto delta_r = p1[R] - p0[R];
+    auto delta_z = p1[Z] - p0[Z];
+    auto tangent = delta_r / delta_z;
+    auto intercept = p0[Z] - p0[R] * delta_z / delta_r;
+
+    // The tangent value given to ConeZ must be positive. However, since
+    // ConeZ creates a double-sheeted cone, the negative cone will be
+    // properly produced as well.
+    return ConeZ{Real3{0, 0, intercept}, std::abs(tangent)};
```

Comment:
```suggestion
    // ConeZ creates a double-sheeted cone, the negative cone will be
    // properly produced as well.
    return ConeZ{Real3{0, 0, intercept}, tangent};
```
The positivity condition was removed somewhat recently (#1832); the restriction now is the tangent is nonzero (and not NaN).